File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.util.*;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[][] matrix = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                matrix[i][j] = sc.nextInt();
            }
        }
        sc.close();
        int[] totalCounts = countColors(matrix, 0, 0, n);
        System.out.println(totalCounts[0]);
        System.out.println(totalCounts[1]);
    }
    
    private static String checkColor(int[][] matrix, int x, int y, int size) {
        int firstColor = matrix[x][y];
        
        for(int i = x; i < x+size; i++) {
            for(int j = y; j < y+size; j++) {
                if (matrix[i][j] != firstColor)
                    return "mixed";
            }
        }
        
        if (firstColor == 0)
            return "white";
        else
            return "blue";
    }
    
    private static int[] countColors(int[][] matrix, int x, int y, int size) {
        String state = checkColor(matrix, x, y, size);
        
        if (state.equals("white"))
            return new int[]{1, 0};
        else if (state.equals("blue"))
            return new int[]{0, 1};
        
        int half = size / 2;
        int[] counts1 = countColors(matrix, x, y, half);
        int[] counts2 = countColors(matrix, x, y+half, half);
        int[] counts3 = countColors(matrix, x+half, y, half);
        int[] counts4 = countColors(matrix, x+half, y+half, half);
        
        int totalWhite = counts1[0] + counts2[0] + counts3[0] + counts4[0];
        int totalBlue = counts1[1] + counts2[1] + counts3[1] + counts4[1];
        return new int[]{totalWhite, totalBlue};
    }
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
import sys

def check_colors(matrix, x, y, size):
    first_color = matrix[x][y]

    for i in range(x, x+size):
        for j in range(y, y+size):
            if matrix[i][j] != first_color:
                return 'mixed'

    if first_color == 0:
        return 'white'
    else:
        return 'blue'

def count_colors(matrix, x, y, size):
    state = check_colors(matrix, x, y, size)

    if state == 'white':
        return (1, 0)
    elif state == 'blue':
        return (0, 1)

    half = size // 2
    w1, b1 = count_colors(matrix, x, y, half)
    w2, b2 = count_colors(matrix, x, y+half, half)
    w3, b3 = count_colors(matrix, x+half, y, half)
    w4, b4 = count_colors(matrix, x+half, y+half, half)
    total_white = w1 + w2 + w3 + w4
    total_blue = b1 + b2 + b3 + b4
    return (total_white, total_blue)

n = int(sys.stdin.readline())
matrix = []
for _ in range(n):
    row = list(map(int, sys.stdin.readline().split()))
    matrix.append(row)

total_white_counts, total_blue_counts = count_colors(matrix, 0, 0, n)
print(total_white_counts)
print(total_blue_counts)